File: src/main/java/com/langleon/dsobuildsim/dto/ItemDefinitionDTO.java
package com.langleon.dsobuildsim.dto;

import com.langleon.dsobuildsim.common.StatType;
import com.langleon.dsobuildsim.items.core.ItemType;
import com.langleon.dsobuildsim.items.core.enums.ItemCategory;
import com.langleon.dsobuildsim.items.core.enums.ItemSlotType;
import com.langleon.dsobuildsim.sets.SetType;

import java.util.List;
import java.util.Map;
import java.util.Set;

public record ItemDefinitionDTO(
        ItemCategory itemCategory,
        ItemType itemType,
        String name,
        int defaultLevel,
        int tier,
        ItemSlotType itemSlotType,
        Map<StatType, Double> rawBaseValues,

        //optional fields depending on category
        Map<StatType, Double> uniqueBaseValues,
        Map<StatType, Double> uniqueRelativeValues,
        Map<StatType, Double> uniqueAbsoluteValues,
        Set<EnchantmentDTO> uniqueEnchantments,
        String uniqueDescription,

        SetType set
        ) {
}

File: frontend/src/app/models/gamedataDTOs/ItemDefinitionDTO.ts
import { EnchantmentDTO } from "./EnchantmentDTO";
import { StatType } from "../../enums/StatType";
import { ItemSlotType } from "../../enums/ItemSlotType";

export interface ItemDefinitionDTO {
    itemCategory: string;
    itemType: string;
    name: string;
    defaultLevel: number;
    tier: number;
    itemSlotType: ItemSlotType;
    rawBaseValues: Record<StatType, number>;
    //optional fields, might be empty
    uniqueBaseValues: Record<StatType, number>;
    uniqueRelativeValues: Record<StatType, number>;
    uniqueAbsoluteValues: Record<StatType, number>;
    uniqueEnchantments: EnchantmentDTO[];
    uniqueDescription: string;
    set: string;
}